### PR TITLE
Making PATH modification optional

### PR DESCRIPTION
**Pull request by Mzack9999**
- Optional path modification
- Supports set/unset via -sp/-up

### CODE REVIEW

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]
@Mzack9999 , Printing all env variable values does not seem necessary/good idea. 

```console
./pdtm 

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.3

		projectdiscovery.io

[INF] [OS: DARWIN] [ARCH: ARM64] [GO: 1.19.4]
[INF] Path to download project binary: /Users/tarun/.pdtm/go/bin

[INF] Path /Users/tarun/.pdtm/go/bin configured in environment variable $PATH: /Users/tarun/.pdtm/go/bin,/opt/homebrew/opt/libpcap/bin,/Users/tarun/.projectdiscovery,/opt/homebrew/opt/openjdk/bin,/Users/tarun/go/bin,/opt/local/bin,/opt/local/sbin,/opt/homebrew/sbin,/opt/homebrew/bin
1. subfinder (latest) (2.5.5)
2. dnsx (latest) (1.1.1)
3. naabu (latest) (2.1.1)
4. httpx (latest) (1.2.6)
5. nuclei (latest) (2.8.8)
6. uncover (latest) (1.0.2)
7. cloudlist (latest) (1.0.2)
8. proxify (latest) (0.0.8)
9. tlsx (latest) (1.0.5)
10. notify (latest) (1.0.4)
11. chaos-client (latest) (0.4.0)
12. shuffledns (latest) (1.0.8)
13. mapcidr (latest) (1.0.3)
14. interactsh-client (latest) (1.0.7)
15. katana (latest) (0.0.3)

```

**Comment by tarunKoyalwar** [comment]
@Mzack9999 , I was thinking we could something similar to this
```console
./pdtm 

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.3

		projectdiscovery.io

[INF] [OS: DARWIN] [ARCH: ARM64] [GO: 1.19.4]
[INF] Path to download project binary: /Users/tarun/.pdtm/go/bin
[INF] Path /Users/tarun/.pdtm/go/bin configured in environment variable $PATH: 

1. subfinder (latest) (2.5.5)
2. dnsx (latest) (1.1.1)
3. naabu (latest) (2.1.1)
4. httpx (latest) (1.2.6)
5. nuclei (latest) (2.8.8)
6. uncover (latest) (1.0.2)
7. cloudlist (latest) (1.0.2)
8. proxify (latest) (0.0.8)
9. tlsx (latest) (1.0.5)
10. notify (latest) (1.0.4)
11. chaos-client (latest) (0.4.0)
12. shuffledns (latest) (1.0.8)
13. mapcidr (latest) (1.0.3)
14. interactsh-client (latest) (1.0.7)
15. katana (latest) (0.0.3)

```

**Review by tarunKoyalwar — APPROVED** [review]
lgtm !

```console
$ ./pdtm 

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.3

		projectdiscovery.io

[INF] [OS: DARWIN] [ARCH: ARM64] [GO: 1.19.4]
[INF] Path to download project binary: /Users/tarun/.pdtm/go/bin
[INF] Path /Users/tarun/.pdtm/go/bin configured in environment variable $PATH
1. subfinder version not found
2. dnsx version not found
3. naabu version not found
4. httpx (latest) (1.2.7)
5. nuclei version not found
6. uncover version not found
7. cloudlist version not found
8. proxify version not found
9. tlsx (latest) (1.0.5)
10. notify version not found
11. chaos-client version not found
12. shuffledns version not found
13. mapcidr version not found
14. interactsh-client version not found
15. katana (latest) (0.0.3)
```

@Mzack9999  since we only track/show/update tools installed via pdtm (i.e $HOME/.pdtm) Instead of printing `nuclei version not found` what do you think if we did something like this

```console
$ ./pdtm 

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.3

		projectdiscovery.io

[INF] [OS: DARWIN] [ARCH: ARM64] [GO: 1.19.4]
[INF] Path to download project binary: /Users/tarun/.pdtm/go/bin
[INF] Path /Users/tarun/.pdtm/go/bin configured in environment variable $PATH

1. httpx (latest) (1.2.7)
2. tlsx (latest) (1.0.5)
3. katana (latest) (0.0.3)

[INF] Tools not tracked[installed] by pdtm

1. nuclei version not found
2. uncover version not found
3. cloudlist version not found
4. notify version not found
5. chaos-client version not found
6. shuffledns version not found
7. mapcidr version not found
8. interactsh-client version not found
9. subfinder version not found
10. dnsx version not found
11. naabu version not found
```

**Comment by tarunKoyalwar** [comment]
since there were some issues with pdtm many users installed tools via `go` tool adding such details/description would help and also avoid conflicts with old installations etc

**Comment by tarunKoyalwar** [comment]
@Mzack9999 , since path modification is disabled by default what do you think of alias ?
i.e when we install say nuclei we alias full path with with `name` this way there won't be any direct path modification or related concerns . 
```console
alias nuclei="/Users/tarun/.pdtm/go/bin/nuclei"
```

What do you think ?

**Comment by Mzack9999** [comment]
Still, to make it persistent, it's necessary to store it somewhere on the system. I think we can suggest to the user the command to execute (pdtm or system commands). As a reference, homebrew indicates to the user the commands to copy/paste and execute.

**Comment by Mzack9999** [comment]
Todo: (cc @tarunKoyalwar )
- Path modification should be enabled (actual default behavior and expected one by users with no setup effort - even if a bit invasive)
- CRUD operations for tools within `$home` directory
- Read-only operations for tools outside `$home` directory

File: internal/runner/runner.go [hunk 1]
```diff
@@ -76,6 +94,10 @@ func (r *Runner) Run() error {
 		}
 	}
 	for _, tool := range r.options.Update {
+		if !path.IsSubPath(homeDir, r.options.Path) {
+			gologger.Error().Msgf("skipping update outside home folder: %s", tool)
+			continue
+		}
```
Comment (tarunKoyalwar) [inline internal/runner/runner.go:100]: I think logic here is different `CRUD operations for tools within $home directory`
@Mzack9999  shouldn't pdtm check path of `nuclei(or other tool)` when updating ?

File: internal/runner/runner.go [hunk 2]
```diff
@@ -27,6 +28,19 @@ func NewRunner(options *Options) (*Runner, error) {
 
 // Run the instance
 func (r *Runner) Run() error {
+	// add default path to $PATH
+	if r.options.SetPath || r.options.Path == defaultPath {
+		if err := path.SetENV(r.options.Path); err != nil {
+			return errorutil.NewWithErr(err).Msgf(`Failed to set path: %s. Add %s to $PATH and run again`)
+		}
+	}
+
```
Comment (tarunKoyalwar) [inline internal/runner/runner.go:37]: while this logic is good . I think we can add some comments to point out that we modify path by default (I think it is  implied but not directly mentioned but would help to understand code if it was explicitly mentioned)

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]
```console
$  go run . -u nuclei

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.4

		projectdiscovery.io

parent is /Users/tarun
sub is /Users/tarun/.pdtm/go/bin
[INF] nuclei: not found in path /Users/tarun/.pdtm/go/bin/nuclei
got error here
[ERR] error while updating nuclei: tool nuclei not found in path /Users/tarun/.pdtm/go/bin/nuclei: skipping
```

### Revisting CRUD operations other than (default PDTM PATH)
- Now after trying this out , I think we don't need all CRUD operations supported for cases when path is different from default PATH

```console
As a User, I think pdtm should detect all pd tools , install them if explicitly mentioned and update my tools (installed with pdtm and without pdtm [Keeping HOME FOLDER restriction HERE]) also I don't want pdtm to take complete ownership (DELETE PERM) of tools that were not installed via pdtm
```
so I think we should only support `READ UPDATE` for tools outside of pdtm default path . this  way we are less invasive keeping User experience same . What do you think @Mzack9999

**Comment by Mzack9999** [comment]
@tarunKoyalwar I'm unable to reproduce it:
```console
$ ls /Users/marcornvh/.pdtm/go/bin/nuclei 
/Users/marcornvh/.pdtm/go/bin/nuclei
$  go run . -u nuclei

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.4

                projectdiscovery.io

[INF] nuclei: already up to date
```
Testing on OSX latest

File: internal/runner/runner.go (hunk 1 above)
Comment (Mzack9999) [reply internal/runner/runner.go:100]: I think here we are doing exactly this, if the provided path is not within the home folder, we skip update. Same happens for install

File: internal/runner/runner.go (hunk 2 above)
Comment (Mzack9999) [reply internal/runner/runner.go:37]: I've updated the readme

**Comment by tarunKoyalwar** [comment]
@Mzack9999  
- currently if I try to update tool not installed by pdtm (or not installed at all) it tries to update it and fails . we probably should add a check here before updating
- Another change related to update that if user wants to update `nuclei` but it is not installed via pdtm (maybe installed via package manager or go install) we can just install a new version in `.pdtm/go/bin` and since PATH is added by default `nuclei` at `.pdtm/go/bin` is picked everytime and this won't be invasive at all

**Review by ehsandeep — APPROVED** [review]
```console
pdtm $./pdtm -ua

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.4

		projectdiscovery.io

[INF] Current pdtm version v0.0.4 (latest)
[INF] subfinder: already up to date
[INF] dnsx: already up to date
[INF] naabu: already up to date
[INF] httpx: already up to date
[INF] nuclei: already up to date
[INF] uncover: already up to date
[INF] cloudlist: already up to date
[INF] proxify: already up to date
[INF] tlsx: already up to date
[INF] notify: already up to date
[INF] chaos-client: already up to date
[INF] shuffledns: already up to date
[INF] mapcidr: already up to date
[INF] interactsh-client: already up to date
[INF] katana: already up to date
[INF] pdtm: already up to date
[INF] asnmap: already up to date


pdtm $./pdtm -ia

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.4

		projectdiscovery.io

[INF] Current pdtm version v0.0.4 (latest)
[INF] subfinder: already installed
[INF] dnsx: already installed
[INF] naabu: already installed
[INF] httpx: already installed
[INF] nuclei: already installed
[INF] uncover: already installed
[INF] cloudlist: already installed
[INF] proxify: already installed
[INF] tlsx: already installed
[INF] notify: already installed
[INF] chaos-client: already installed
[INF] shuffledns: already installed
[INF] mapcidr: already installed
[INF] interactsh-client: already installed
[INF] katana: already installed
[INF] pdtm: already installed
[INF] asnmap: already installed
```

**Review by tarunKoyalwar — APPROVED** [review]
lgtm !

```console
$ ./pdtm 

                ____          
     ____  ____/ / /_____ ___ 
    / __ \/ __  / __/ __ __  \
   / /_/ / /_/ / /_/ / / / / /
  / .___/\__,_/\__/_/ /_/ /_/ 
 /_/                          v0.0.4

		projectdiscovery.io

[INF] Current pdtm version v0.0.4 (latest)
[INF] [OS: DARWIN] [ARCH: ARM64] [GO: 1.20.2]
[INF] Path to download project binary: /Users/tarun/.pdtm/go/bin
[INF] Path /Users/tarun/.pdtm/go/bin configured in environment variable $PATH
1. subfinder (latest) (2.5.7)
2. dnsx (latest) (1.1.3)
3. naabu (latest) (2.1.4)
4. httpx (latest) (1.2.9)
5. nuclei (not installed)
6. uncover (latest) (1.0.3)
7. cloudlist (latest) (1.0.2)
8. proxify (latest) (0.0.8)
9. tlsx (latest) (1.0.6)
10. notify (latest) (1.0.4)
11. chaos-client (latest) (0.5.1)
12. shuffledns (latest) (1.0.8)
13. mapcidr (latest) (1.1.1)
14. interactsh-client (latest) (1.1.2)
15. katana (latest) (1.0.0)
16. pdtm (latest) (0.0.4)
17. asnmap (latest) (1.0.2)

```

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**